src/core_plugins/input_control_vis/public/components/editor/controls_tab.js

```diff
@@ -0,0 +1,176 @@
+      type: 'index-pattern',
+      fields: ['title'],
+      search: `${search}*`,
+      search_fields: ['title'],
+      perPage: 100
+    });
+    return resp.savedObjects;
+  }
+
+  async getIndexPattern(indexPatternId) {
+    return await this.props.scope.vis.API.indexPatterns.get(indexPatternId);
+  }
+
+  setVisParam(paramName, paramValue) {
+    const params = _.cloneDeep(this.props.scope.vis.params);
+    params[paramName] = paramValue;
+    this.props.stageEditorParams(params);
+  }
+
+  handleLabelChange(controlIndex, evt) {
+    const updatedControl = this.props.scope.vis.params.controls[controlIndex];
+    updatedControl.label = evt.target.value;
+    this.setVisParam('controls', setControl(this.props.scope.vis.params.controls, controlIndex, updatedControl));
+  }
+
+  handleIndexPatternChange(controlIndex, evt) {
+    const updatedControl = this.props.scope.vis.params.controls[controlIndex];
+    updatedControl.indexPattern = evt.value;
+    updatedControl.fieldName = '';
+    this.setVisParam('controls', setControl(this.props.scope.vis.params.controls, controlIndex, updatedControl));
+  }
+
+  handleFieldNameChange(controlIndex, evt) {
+    const updatedControl = this.props.scope.vis.params.controls[controlIndex];
+    updatedControl.fieldName = evt.value;
+    this.setVisParam('controls', setControl(this.props.scope.vis.params.controls, controlIndex, updatedControl));
+  }
+
+  handleCheckboxOptionChange(controlIndex, optionName, evt) {
+    const updatedControl = this.props.scope.vis.params.controls[controlIndex];
+    updatedControl.options[optionName] = evt.target.checked;
+    this.setVisParam('controls', setControl(this.props.scope.vis.params.controls, controlIndex, updatedControl));
+  }
+
+  handleNumberOptionChange(controlIndex, optionName, evt) {
+    const updatedControl = this.props.scope.vis.params.controls[controlIndex];
+    updatedControl.options[optionName] = parseFloat(evt.target.value);
+    this.setVisParam('controls', setControl(this.props.scope.vis.params.controls, controlIndex, updatedControl));
+  }
+
+  handleRemoveControl(controlIndex) {
+    this.setVisParam('controls', removeControl(this.props.scope.vis.params.controls, controlIndex));
+  }
+
+  moveControl(controlIndex, direction) {
+    this.setVisParam('controls', moveControl(this.props.scope.vis.params.controls, controlIndex, direction));
+  }
+
+  handleAddControl() {
+    this.setVisParam('controls', addControl(this.props.scope.vis.params.controls, newControl(this.state.type)));
+  }
+
+  renderControls() {
+    return this.props.scope.vis.params.controls.map((controlParams, controlIndex) => {
+      let controlEditor = null;
+      switch (controlParams.type) {
+        case 'list':
+          controlEditor = (
+            <ListControlEditor
+              controlIndex={controlIndex}
+              controlParams={controlParams}
+              handleIndexPatternChange={this.handleIndexPatternChange.bind(this, controlIndex)}
```

Comment:
These binds too should be moved to constructor or es7 syntax (and below)

Reply:
Question for @weltenwort and @kjbekkelund (and perhaps @cjcenizal when you get back) - 

This is actually a more interesting example because it's not just a simple bind but binds the value.  You could resolve this by passing `controlIndex` into `handleIndexPatternChange` inside the `ListControlEditor` , `RangeControlEditor`, and `ControlEditor ` components.  So something like:

```
export function ListControlEditor(props) {
  function handleIndexPatternChange() {
    props.handleIndexPatternChange(props.controlIndex);
  }
   return (
     <div>
       <IndexPatternSelect
         value={props.controlParams.indexPattern}
         onChange={handleIndexPatternChange}
         getIndexPatterns={props.getIndexPatterns}
      />
....
```

which resolves the performance issues of binding in render (Well, I think, I'm a little confused about how to do that with a functional component.  I'm not 100% sure my version above does not become the equivalent of binding in render if it was using class syntax).  But then you have all this extra code for every specialized type of component.

Have you guys come across this before in your React travels and have any words of wisdom, or preferences for how you would handle this?  Ignore the performance implications and just do the easy bind in render? Or write extra code to avoid it? Or something else I haven't thought of?

Reply:
Unfortunately your solution would have the same effect as using `.bind()` in that both create a new function on every render call. I can think of a few options:

* cache the handlers with `controlIndex` partially applied (I did this with [recompose's `withHandlers` HOC](https://github.com/acdlite/recompose/blob/master/docs/API.md#withhandlers) in other projects, but I think we don't yet have that incredibly useful dependency yet :wink: could be implemented in the stateful component via `ReceiveProps`)
* expect the `ListControlEditor` to supply its `controlIndex` when calling its `props.handleIndexPatternChange` (this couples the components a bit stronger)
* somehow store the index in the doc and retrieve it from the `event.target` (messy, would consider this the last options)
* just accept that it creates new functions (if the component is pure and rarely re-rendered I wouldn't expect this to cause problems)

Reply:
Nevermind @weltenwort and @kjbekkelund  (and @cjcenizal) - @nreese is a step ahead of me and came up with an elegant solution.

Reply:
nice, what does it look like?